File: leetcode-scala/src/leetcode/medium/IncreasingTripletSubsequence.scala
package leetcode.medium

import scala.annotation.tailrec

object IncreasingTripletSubsequence {
  def increasingTriplet(nums: Array[Int]): Boolean = {
    var fi = Int.MaxValue
    var se = Int.MaxValue
    for (i <- nums.indices) {
      nums.apply(i) match {
        case tmp: Int if(nums.apply(i) <= fi) => fi = tmp
        case tmp: Int if(nums.apply(i) <= se) => se = tmp
        case (_) => return true
      }
    }
    false
  }

}

File: leetcode-java/src/leetcode/patterns/IncreasingTripletSubsequence.java
package leetcode.patterns;

public class IncreasingTripletSubsequence {

    public static boolean increasingTriplet(int[] nums) {
        var m1 = Integer.MAX_VALUE;
        var m2 = Integer.MAX_VALUE;
        for (var num: nums) {
            if (m1 > num) {
                m1 = num;
            } else if (m2 > num) {
                m2 = num;
            } else {
                return true;
            }
        }
        return false;
    }

    public static void main(String[] args) {
        increasingTriplet(new int[]{5, 4, 3, 2, 1});
    }
}
